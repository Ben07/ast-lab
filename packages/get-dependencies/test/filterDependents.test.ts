import path from 'path';
import resolve from 'resolve';
// @ts-ignore
import { enumerateArgsTestFunction, configArgs } from 'lazy-jest';
import _filterDependents from '../src/filterDependents';

const FIXTURE_DIR = 'test/__fixtures__';
const resolvePath = (file: string) => path.resolve(FIXTURE_DIR, file);

function filterDependents(
  basedir: string,
  sources: string[],
  imports: {
    [mod: string]: string[]
  },
  ...args: any[]
) {
  return _filterDependents(
    sources.map(s => resolvePath(path.join(basedir, s))),
    makeTargets(basedir, imports),
    ...args
  ).map(absolute => absolute.replace(__dirname, ''));
}
function makeTargets(basedir: string, data: {
  [mod: string]: string[]
}) {
  return new Map( Object.keys(data).map(mod => {
    return [
      resolve.sync(mod, { basedir: path.resolve(FIXTURE_DIR, basedir) }),
      new Set(data[mod])
    ];
  }));
}

describe('filterDependents()', () => {
  enumerateArgsTestFunction(
    filterDependents.bind(null, 'imports'),
    configArgs()
    .arg('sources', [
      ['base.js'],
      ['importAlias.js'],
      ['importAll.js'],
      ['importDefault.js'],
      ['importNamed.js'],
      ['noImports.js']
    ])
    .arg('targets', [
      { './base': ['a', 'default'] }
    ]),
    'imports'
  );

  enumerateArgsTestFunction(
    filterDependents.bind(null, 'exports'),
    configArgs()
    .arg('sources', [
      ['exportAllFrom.js'],
      ['exportDefault.js'],
      ['exportDefaultAliasFrom.js'],
      ['exportNamedFrom.js'],
      ['noExports.js']
    ])
    .arg('targets', [
      { './exportNamed': ['a'] },
      { './exportDefault': ['default'] }
    ]),
    'exports from'
  );

  enumerateArgsTestFunction(
    filterDependents.bind(null, 'module-alias'),
    configArgs()
    .arg('sources', [
      ['unrelated-folder.js'],
      ['unrelated-name.js'],
    ])
    .arg('targets', [
      { './a': ['default'] }
    ])
    .arg('options', [{
      alias: {
        'alias-folder': './',
        'alias-a': './a'
      }
    }]),
    'module alias'
  );
  
  enumerateArgsTestFunction(
    filterDependents.bind(null, 'module-directory'),
    configArgs()
    .arg('sources', [
      ['a.js', 'b.js'],
    ])
    .arg('targets', [
      { './a': ['default'] }
    ])
    .arg('options', [{
      moduleDirectory: ['../']
    }]),
    'module alias'
  );

  enumerateArgsTestFunction(
    filterDependents.bind(null, 'circular'),
    configArgs()
    .arg('sources', [
      ['a.js', 'b.js', 'c.js'],
    ])
    .arg('targets', [
      { './a': ['default'] }
    ])
    .arg('options', [{
      moduleDirectory: ['../']
    }]),
    'circular'
  );
});
