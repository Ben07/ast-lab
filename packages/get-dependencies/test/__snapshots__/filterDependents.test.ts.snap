// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterDependents() exports from - (["sources", "targets"]) - [["exportAllFrom.js"],{"./exportDefault":["default"]}] 1`] = `Array []`;

exports[`filterDependents() exports from - (["sources", "targets"]) - [["exportAllFrom.js"],{"./exportNamed":["a"]}] 1`] = `
Array [
  "/__fixtures__/exports/exportAllFrom.js",
]
`;

exports[`filterDependents() exports from - (["sources", "targets"]) - [["exportDefault.js"],{"./exportDefault":["default"]}] 1`] = `
Array [
  "/__fixtures__/exports/exportDefault.js",
]
`;

exports[`filterDependents() exports from - (["sources", "targets"]) - [["exportDefault.js"],{"./exportNamed":["a"]}] 1`] = `Array []`;

exports[`filterDependents() exports from - (["sources", "targets"]) - [["exportDefaultAliasFrom.js"],{"./exportDefault":["default"]}] 1`] = `
Array [
  "/__fixtures__/exports/exportDefaultAliasFrom.js",
]
`;

exports[`filterDependents() exports from - (["sources", "targets"]) - [["exportDefaultAliasFrom.js"],{"./exportNamed":["a"]}] 1`] = `Array []`;

exports[`filterDependents() exports from - (["sources", "targets"]) - [["exportNamedFrom.js"],{"./exportDefault":["default"]}] 1`] = `Array []`;

exports[`filterDependents() exports from - (["sources", "targets"]) - [["exportNamedFrom.js"],{"./exportNamed":["a"]}] 1`] = `
Array [
  "/__fixtures__/exports/exportNamedFrom.js",
]
`;

exports[`filterDependents() exports from - (["sources", "targets"]) - [["noExports.js"],{"./exportDefault":["default"]}] 1`] = `Array []`;

exports[`filterDependents() exports from - (["sources", "targets"]) - [["noExports.js"],{"./exportNamed":["a"]}] 1`] = `Array []`;

exports[`filterDependents() imports - (["sources", "targets"]) - [["base.js"],{"./base":["a","default"]}] 1`] = `
Array [
  "/__fixtures__/imports/base.js",
]
`;

exports[`filterDependents() imports - (["sources", "targets"]) - [["importAlias.js"],{"./base":["a","default"]}] 1`] = `
Array [
  "/__fixtures__/imports/importAlias.js",
]
`;

exports[`filterDependents() imports - (["sources", "targets"]) - [["importAll.js"],{"./base":["a","default"]}] 1`] = `
Array [
  "/__fixtures__/imports/importAll.js",
]
`;

exports[`filterDependents() imports - (["sources", "targets"]) - [["importDefault.js"],{"./base":["a","default"]}] 1`] = `
Array [
  "/__fixtures__/imports/importDefault.js",
]
`;

exports[`filterDependents() imports - (["sources", "targets"]) - [["importNamed.js"],{"./base":["a","default"]}] 1`] = `
Array [
  "/__fixtures__/imports/importNamed.js",
]
`;

exports[`filterDependents() imports - (["sources", "targets"]) - [["noImports.js"],{"./base":["a","default"]}] 1`] = `Array []`;
